src/vision/status_quo/barbara_plays_with_async.md

```diff
@@ -0,0 +1,387 @@
+
+   = note: expected closure `[closure@src\main.rs:16:44: 19:14]`
+              found closure `[closure@src\main.rs:21:44: 24:14]`
+   = note: no two closures, even if identical, have the same type
+   = help: consider boxing your closure and/or using it as a trait object
+```
+
+This error doesn't actually surprise Barbara that much, as she is familiar with
+the idea of having to box objects sometimes.  She does
+notice the "consider boxing your closure" error, but thinks that this is not
+likely the correct solution.  Instead, she thinks that she should box the
+entire future.
+
+She first adds explicit type annotations to the Vec:
+
+```rust
+let mut futures: Vec<Box<dyn Future<Output=()>>> = Vec::new();
+```
+
+She then notices that her IDE (VSCode + rust-analyzer) has a new error on
+each call to push.  The code assist on each error says `Store this in the heap
+by calling 'Box::new'`.  She is exactly what she wants, and it happy that
+rust-analzyer perfectly handled this case.
+
+Now each future is boxed up, ut there is one final error yet,
+this time on the call to `join_all(futures).await`:
+
+```
+error[E0277]: `dyn futures::Future<Output = ()>` cannot be unpinned
+  --> src\main.rs:34:31
+   |
+34 |     futures::future::join_all(futures).await;
+```
+
+Barbara has been around rust for long enough to know that there is a `Box::pin`
+API, but she doesn't really understand what it does, nor does she have a good
+intuition about what this API is for.  But she is accustomed to just trying
+things in rust to see if they work.  And indeed, after changing `Box::new` to
+`Box::pin` (and adjusting the type of the Vec), the code compiles and runs
+successfully.
+
+But even though the run is working correctly, she wishes she had a better idea
+why pinning is necessary here and feels a little uneasy having to use something
+she doesn't yet understand well.
+
+As one final task, Barbara wants to try to replace the chained call to `then`
+with a async block.  She remembers that these were a big deal in a recent
+release of rust, and that they looked a lot nicer than a long chain of `then`
+calls.  She doesn't remember the exact syntax for this, but she read a blog
+post about async rust a few weeks ago, and has a vague idea of how it looks.
+
+She tries writing this:
+
+```rust
+futures.push(Box::pin(async || {
+    tokio::time::sleep(Duration::from_millis(delay)).await;
+    println!("Done after {}ms", delay);
+}));
+```
+
+The compiler gives an error:
+
+```
+error[E0658]: async closures are unstable
+  --> src\main.rs:14:31
+   |
+14 |         futures.push(Box::pin(async || {
+   |                               ^^^^^
+   |
+   = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
+   = help: add `#![feature(async_closure)]` to the crate attributes to enable
+   = help: to use an async block, remove the `||`: `async {`
+```
+
+Barbara knows that async are stable and using this nightly feature isn't what
```

Comment:
```suggestion
Barbara knows that async isn't stable and using this nightly feature isn't what
```

Reply:
Actually, what I wanted to say is:  "Barbara knows that async *is* stable".  Do I have that right?